### PR TITLE
Better traceback for bad imports in custom path

### PR DESCRIPTION
**Pull request by dset0x**

### CODE REVIEW

File: werkzeug/utils.py [hunk 1]
```diff
@@ -423,13 +423,7 @@ def import_string(import_name, silent=False):
             return sys.modules[import_name]
 
         module_name, obj_name = import_name.rsplit('.', 1)
-        try:
-            module = __import__(module_name, None, None, [obj_name])
-        except ImportError:
-            # support importing modules not yet set up by the parent module
-            # (or package for that matter)
-            module = import_string(module_name)
```
Comment (untitaker) [inline werkzeug/utils.py]: The tests don't fail because of your PR, but I think this line is there for a reason (no tests there for it though?)

Comment (dset0x) [inline werkzeug/utils.py]: This also made me raise an eyebrow, but I could not understand what "not yet set up" implied either. However, the case that I introduced in the tests, does cause the `ImportError` of line 428 to be raised.

FWIW I use this patch with https://github.com/inveniosoftware/flask-registry and had no obvious issues.

Comment (untitaker) [inline werkzeug/utils.py]: I can't find a way to break your code either. The original code dates back to https://github.com/mitsuhiko/werkzeug/commit/9aa53b8a8a7203bcbaf62d2ed1a12c8d7e54d3ce, but there's insufficient explanation there.

Comment (untitaker) [inline werkzeug/utils.py]: I'll assign this to @mitsuhiko, maybe he (still) can come up with a testcase for the linked commit.

**Comment by RonnyPfannschmidt** [comment]
i suspect a potential use case lingering in lazy code loaders that inject modules into sys.modules at their own import

on closer examination the internal logic is grown a bit strange and should probably do a different loop altogether

**Comment by davidism** [comment]
We can preserve the "not set up by parent" case by saving the exception and raising it in the `AttributeError` block. However, this just adds more complexity to a code path no one can reproduce. And it makes the chained traceback in Python 3 look like it's coming from the wrong place. I'm inclined to merge this as-is, leaving this as a note in case someone comes up with a failing test case. 

```diff
diff --git a/werkzeug/utils.py b/werkzeug/utils.py
index 5c8f5125..b3e859f7 100644
--- a/werkzeug/utils.py
+++ b/werkzeug/utils.py
@@ -423,10 +423,17 @@ def import_string(import_name, silent=False):
             return sys.modules[import_name]
 
         module_name, obj_name = import_name.rsplit('.', 1)
-        module = __import__(module_name, globals(), locals(), [obj_name])
+        preserved_import_error = None
+        try:
+            module = __import__(module_name, globals(), locals(), [obj_name])
+        except ImportError as e:
+            preserved_import_error = e
+            module = import_string(module_name)
         try:
             return getattr(module, obj_name)
         except AttributeError as e:
+            if preserved_import_error is not None:
+                raise preserved_import_error
             raise ImportError(e)
 
     except ImportError as e:
```

**Comment by dset0x** [comment]
Nearly three and a half years later, I still vividly recall the dozens of hours of pain that this was helping solve for the (now-deprecated) referenced module. At any rate, thanks for the merge. I hope it helps.

**Comment by davidism** [comment]
Sorry it took so long :disappointed: It came up in a much more recent issue, so it will still help. Thanks for working on it.

**Pull request merged**